libs/datamodel/core/src/transform/ast_to_dml/db/relations.rs

```diff
@@ -0,0 +1,219 @@
+use super::{context::Context, types::RelationField};
+use crate::ast;
+use std::collections::BTreeSet;
+
+#[derive(PartialOrd, Ord, PartialEq, Eq, Debug, Clone)]
+enum OneToManyRelationFields {
+    Forward(ast::FieldId),
+    Back(ast::FieldId),
+    Both(ast::FieldId, ast::FieldId),
+}
+
+#[derive(PartialOrd, Ord, PartialEq, Eq, Debug, Clone)]
+enum RelationType {
+    ImplicitManyToMany {
+        field_a: ast::FieldId,
+        field_b: ast::FieldId,
+    },
+    OneToOne {
+        field_a: ast::FieldId,
+        field_b: ast::FieldId,
+    },
+    OneToMany(OneToManyRelationFields),
+}
+
+#[derive(PartialOrd, Ord, PartialEq, Eq, Debug, Clone)]
+pub(super) struct Relation<'ast> {
+    /// The `name` argument in `@relation`.
+    relation_name: Option<&'ast str>,
+    tpe: RelationType,
+}
+
+#[derive(Debug, Default)]
+pub(super) struct Relations<'ast> {
+    // A relation is always between two models. One model is assigned the role
+    // of "model A", and the other is "model B". The meaning of "model A" and
+    // "model B" depends on the type of relation.
+    //
+    // - In implicit many-to-many relations, model A and model B are ordered
+    //   lexicographically, by model name, and failing that by relation field
+    //   name. This order must be stable in order for the columns in the
+    //   implicit many-to-many relation table columns and the data in them to
+    //   keep their meaning.
+    // - In one-to-one and one-to-many relations, model A is the one carrying
+    //   the referencing information and possible constraint. For example, on a
+    //   SQL database, model A would correspond to the table with the foreign
+    //   key constraint, while model B would correspond to the table referenced
+    //   by the foreign key.
+
+    // Storage
+    relations_storage: Vec<Relation<'ast>>,
+
+    // Indexes for efficient querying.
+    //
+    // Why BTreeSets?
+    //
+    // - We can't use a BTreeMap because there can be more than one relation
+    //   between two models
+    // - We use a BTree because we want range queries. Meaning that with a
+    //   BTreeSet, we can efficiently ask:
+    //   - Give me all the relations on other models that point to this model
+    //   - Give me all the relations on this model that point to other models
+    //
+    // Where "on this model" doesn't mean "the relation field is on the model"
+    // but "the foreigr key is on this model" (= this model is model a)
+
+    // (model_a, model_b, relation_idx)
+    forward: BTreeSet<(ast::ModelId, ast::ModelId, usize)>,
+    // (model_b, model_a, relation_idx)
+    back: BTreeSet<(ast::ModelId, ast::ModelId, usize)>,
+}
+
+impl<'ast> Relations<'ast> {
+    /// Iterator over (referenced_model_id, relation)
+    #[allow(dead_code)] // not used _yet_
+    pub(super) fn relations_from_model(
```

Comment:
So, do I understand this correctly: we get the relations that hold `fields` and `references`, including the referential actions from this method?

Maybe we should, again, have proper rustdocs explaining what exactly are we giving here.

These are very important and useful methods, but we should aim for the best possible documentation on a subject that is, probably, the most complex part of the datamodel crate.

Reply:
These should be used to build walkers, which would in turn be the public API. That's the low level private API that demonstrates we can query the right things.

Reply:
Will add docs.